[Partitioner] Enable more than one partitions assigned to the same device

beicy opened the pull request:
Summary:
The PR allows more than one partitions to be assigned into one device.
This happens when the number of generated DAGNodes is larger than the number of devices for any backendKind . E.g:                                
node1(6GB) -> node2(14GB) -> node3(6GB). The memory limitation is 16GB, and there is only 2 devices. So we assign node1 and node3 into one device.  
Documentation:

Test Plan:
added unnitest. 


#2298 
Please see a detailed explanation of how to fill out the fields in the relevant sections in PULL_REQUEST.md.

opti-mix commented on lib/Partitioner/Partitioner.cpp:
```diff
@@ -26,6 +26,15 @@
+namespace {
+// Sorted the std::pair<DAGNode *, uint64_t> based on the second from min to
+// max.
+auto sortMinMemory = [](const std::pair<DAGNode *, uint64_t> &a,
```
Why is it a closure? It can be just a usual function, or?

beicy replied:
No special reason, just followed some code sample :P  Now removed.

opti-mix commented on lib/Partitioner/Partitioner.cpp:
```diff
@@ -62,12 +71,12 @@ void Partitioner::dumpDAG(llvm::StringRef dotFilename) const {
       auto child = node->children[i];
       DescriptionBuilder db(child->name.c_str());
-      db.addParam("BackendKind", backendName_.at(child->backendKind));
+      std::string backendName = backendMap_.at(child->backendKind).name;
```
`const std::string &backendName`?

opti-mix commented on lib/Partitioner/Partitioner.cpp:
```diff
@@ -492,7 +501,78 @@ NodeToFunctionMap Partitioner::selectPartitions(Function *F,
-void Partitioner::adjustLogicalDeviceID() {}
+llvm::Error Partitioner::adjustLogicalDeviceID(NodeToFunctionMap &mapping) {
+  for (auto &dag : partitions_) {
+    // Traver the DAGNodes, and get list of with each backendKind.
```
Traverse the DAGNodes and get list of ??? with each backendKind.

List of what? It seems like this sentence is not finished or misses some words.

opti-mix commented on lib/Partitioner/Partitioner.cpp:
```diff
@@ -492,7 +501,78 @@ NodeToFunctionMap Partitioner::selectPartitions(Function *F,
+llvm::Error Partitioner::adjustLogicalDeviceID(NodeToFunctionMap &mapping) {
+  for (auto &dag : partitions_) {
+    // Traver the DAGNodes, and get list of with each backendKind.
+    std::map<BackendKind, std::vector<DAGNode *>> backendKindNum;
```
The name `backendKindNum` is a bit odd as you do not map the backendKind to a number. If I understand correctly, you map the backendKind to a set of DAGNodes and then you check the number of elements in this set.

opti-mix commented on lib/Partitioner/Partitioner.cpp:
```diff
@@ -492,7 +501,78 @@ NodeToFunctionMap Partitioner::selectPartitions(Function *F,
+      // Assume we have n devices and m partitions, where m > n. If these m
+      // partitions can be assigned to n devices, there must be 1 device have at
+      // least (m - n + 1) partitions. Therefore, we just need to check the
+      // minial (m - n + 1) partitions can be put into 1 device while the device
```
minial -> minimal

opti-mix commented on lib/Partitioner/Partitioner.cpp:
```diff
@@ -492,7 +501,78 @@ NodeToFunctionMap Partitioner::selectPartitions(Function *F,
+
+      // Assume we have n devices and m partitions, where m > n. If these m
+      // partitions can be assigned to n devices, there must be 1 device have at
+      // least (m - n + 1) partitions. Therefore, we just need to check the
```
> If these m partitions can be assigned to n devices, there must be 1 device have at least (m - n + 1) partitions. 

Is that true? E.g. if we have 10 partitions and 5 devices, you say that at least one device should get (10 - 5 + 1) = 6 partitions. Why? Why can't each device e.g. get two partitions?

beicy replied:
Good catch!  Thanks for pointing this! I thought the putting n+1 items to n boxes issue... I will rewrite this algorithm.

opti-mix commented on tests/unittests/PartitionerTest.cpp:
```diff
@@ -494,29 +497,66 @@ TEST_F(PartitionerTest, SimpleHeterogeneousPartitioning) {
+}
+
+/// Test assigning more than one partitions in to one device.
+TEST_F(PartitionerTest, HeterogeneousPartitioning1) {
```
Please give this test a more meaningful name.

opti-mix replied:
You just added/modified heterogeneous partitioning tests in this PR, but this new feature could be used also with a single type of backends, right? You should test this case as well.

beicy commented:
@opti-mix @gcatron  The updates in this version:
1. renamed `adjustLogicalDeviceID` with `assignLogicalDeviceID`;
2. moved `assignLogicalDeviceID` before "doPartitioning". Now it is the last step of optimization.
3. Added unittest for heterogeneous partition(logicalIDTest1) and single backendKind (logicalIDTest0).
Thanks!

gcatron commented on tests/unittests/PartitionerTest.cpp:
```diff
@@ -530,6 +533,87 @@ TEST_F(PartitionerTest, SimpleHeterogeneousPartitioning) {
+  mod_.clear();
+}
+
+/// Test assigning more than one partitions in to one device.
```
This test has the same comment as the one above, can you make the comments more specific. I'm a little confused what the difference is.

beicy replied:
Done. One test for single backendKind, one for heterogeneous partitioning.

opti-mix commented on lib/Partitioner/Partitioner.cpp:
```diff
@@ -26,6 +26,13 @@
 
+// Sorted the std::pair<DAGNode *, uint64_t> based on the second from min to
+// max.
+auto sortMinMemory = [](const std::pair<Function *, uint64_t> &a,
```
This is just a simple function. No need to use a capture here.

opti-mix commented on include/glow/Partitioner/Partitioner.h:
```diff
@@ -195,10 +225,15 @@ class Partitioner {
+  DeviceIDTy assignLogicalDeviceID(NodeToFunctionMap &partitions);
+
+  /// Check if this partition way \p partitions satisfies number of physical
+  /// devices restriction. I.e. checkif the number of logical devices is less
```
checkif -> check if

opti-mix commented on lib/Partitioner/Partitioner.cpp:
```diff
@@ -62,12 +69,12 @@ void Partitioner::dumpDAG(llvm::StringRef dotFilename) const {
-      db.addParam("BackendKind", backendName_.at(child->backendKind));
+      const std::string &backendName = backendMap_.at(child->backendKind).name;
+      db.addParam("BackendKind", backendName);
       myfile << child->name << " [ label = \"" << escapeDottyString(db) << "\"";
```
Please use escapeDottyString on all names/dynamic strings. I've seen some issues if we don't do it.

gcatron commented on lib/Partitioner/Partitioner.cpp:
```diff
@@ -485,14 +520,124 @@ NodeToFunctionMap Partitioner::selectPartitions(Function *F,
+
+    // Assume we have n devices(NOTE: here the n devices have the same avaiable
+    // memory size, and the following algorithm can find the accurate result. If
+    // the momery size are differnt, this assignment issue will be a NP problem
```
momery ->memory

gcatron commented on lib/Partitioner/Partitioner.cpp:
```diff
@@ -485,14 +520,124 @@ NodeToFunctionMap Partitioner::selectPartitions(Function *F,
+        }
+        usedMem += nodeSize[i].second;
+      }
+      // Step 4 : if k found in step 3 is larger than (m - 1) / n + 1, this
```
larger -> smaller,
Also it might be helpful to call out in the comment that k = i- start.

gcatron commented on lib/Partitioner/Partitioner.cpp:
```diff
@@ -485,14 +520,124 @@ NodeToFunctionMap Partitioner::selectPartitions(Function *F,
+        return logicalDeviceID;
+      }
+
+      // Step 5 : Assign these partitios which are assigned to one device with
```
partitios -> partitions

opti-mix commented on lib/Partitioner/Partitioner.cpp:
```diff
@@ -26,6 +26,13 @@
+auto sortMinMemory(const std::pair<Function *, uint64_t> &a,
+                   const std::pair<Function *, uint64_t> &b) -> bool {
+  return a.second <= b.second;
+};
```
No need for a semicolon.

opti-mix commented on lib/Partitioner/Partitioner.cpp:
```diff
@@ -84,7 +91,8 @@ void Partitioner::dumpDAG(llvm::StringRef dotFilename) const {
     for (size_t j = 0; j < root->children.size(); j++) {
       auto child = root->children[j];
-      myfile << root->name << " -> " << child->name << ";";
+      myfile << escapeDottyString(root->name) << " -> "
```
You probably need to add `\""` before and after each of these `escapeDottyString` calls, so that it produces `"Escaped string"`.

opti-mix commented on lib/Partitioner/Partitioner.cpp:
```diff
@@ -93,19 +101,46 @@ void Partitioner::dumpDAG(llvm::StringRef dotFilename) const {
 }
 
+llvm::Error
+Partitioner::logicalDevicesValidation(NodeToFunctionMap &partitions) {
```
Do you have a Doxygen comment for this function anywhere? If not, please add it.

beicy replied:
Yes. In Partitioner.h.

opti-mix commented on lib/Partitioner/Partitioner.cpp:
```diff
@@ -485,14 +520,125 @@ NodeToFunctionMap Partitioner::selectPartitions(Function *F,
+  // Step 3 : assign each partition with a logical device id. The partitions
+  // with the same logical device id will be assigned into the same physical
+  // device.
+  // logicalDeviceID_ = assignLogicalDeviceID(mapping, logicalDeviceID_);
```
Why is this line commented out? Does it need to be removed?

beicy replied:
Yes, these lines (line 535-526) need to be removed.

bertmaher commented on lib/Partitioner/Partitioner.cpp:
```diff
@@ -26,6 +26,13 @@
 
+// Sorted the std::pair<DAGNode *, uint64_t> based on the second from min to
+// max.
+auto sortMinMemory(const std::pair<Function *, uint64_t> &a,
```
Is there  a reason for using the `auto foo() -> bool` syntax instead of simply `bool foo()`?  It feels kind of distracting to use this style here as a one-off.

opti-mix replied:
Good catch! Initially, I noticed the closure and asked @beicy to change it to a function. But I haven't noticed that she forgot to change the return type while changing it ;-)

beicy replied:
Thanks! I didn't notice that :P Will add a follow up PR.
